spamc.go

```diff
@@ -67,10 +68,148 @@ func (c *Client) send(
 	return conn, nil
 }
 
+// rearrangeMessageReader will rearrange the message reader to make sure SA will understand the message
+func (c Client) rearrangeMessageReader(message io.ReadSeeker, size int64) (io.Reader, int64, error) {
+	var headers bytes.Buffer
+	var rest bytes.Buffer
+
+	isASCII := func(m []byte) bool {
+		for _, c := range m {
+			if c > 127 {
+				return false
+			}
+		}
+		return true
+	}
+
+	r := bufio.NewReader(message)
+
+	var returnError error
+
+	// we assume that there is no newline between headers and body, so increment size
+	size += 2
+
+	for {
+		chars, err := r.Peek(2)
+		if err != nil {
+			returnError = err
+			break
+		}
+		if bytes.Equal(chars, []byte("\r\n")) {
+			// we were wrong there was a newline between headers and body
+			// decrement size
+			size -= 2
+			// read and discard it, and we are done
+			_, err = r.Read(make([]byte, 2))
+			if err != nil {
+				returnError = err
+			}
+			break
+		}
+		if chars[0] == ' ' || chars[0] == '\t' {
+			// if we have a previous key append value, and continue
+			// if not we are not reading headers
+			if headers.Len() > 0 {
+				line, err := r.ReadSlice('\n')
+				if err != nil {
+					rest.Write(line)
+					returnError = err
+					break
+				}
+				if !isASCII(line) {
```

Comment:
Note that email message headers *can* contain non-ASCII characters; see e.g. https://tools.ietf.org/html/rfc6532

Reply:
Hmm thought that these must always be encoded .. and that turns out to be wrong .. 
> The bodies of header fields are allowed
> to contain Unicode characters, but the header field names themselves
> must consist of ASCII characters only
From rfc6532, we should only test the header name not the value, we could just drop the isASCII test it might be better.